### PR TITLE
Refactoring api tests to use notebook data

### PR DESCRIPTION
**Pull request by shanaxel42**
The api test now load and run their corresponding notebooks with the test data then check the intermediate variables.

### CODE REVIEW

File: definitions.py [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import os
```
Comment (ttung) [inline definitions.py]: is this file actually necessary?  can you just do:

```
import starfish
ROOT_DIR = os.path.dirname(os.path.dirname(os.path.abspath(starfish.__file__)))
```

File: starfish/test/full_pipelines/api/test_dartfish.py [hunk 2]
```diff
@@ -1,26 +1,24 @@
...
 
-from starfish import Experiment, IntensityTable
-from starfish.image._filter.scale_by_percentile import ScaleByPercentile
-from starfish.image._filter.zero_by_channel_magnitude import ZeroByChannelMagnitude
-from starfish.spots._detector.pixel_spot_detector import PixelSpotDetector
+from starfish import IntensityTable
 from starfish.types import Features
 
+os.environ["USE_TEST_DATA"] = "1"
+sys.path.append(os.path.join(ROOT_DIR, "notebooks/py/"))
```
Comment (ttung) [inline starfish/test/full_pipelines/api/test_dartfish.py]: this should be `os.path.join(ROOT_DIR, "notebooks", "py")` in case someone is on Windows.

File: starfish/test/full_pipelines/api/test_iss_api.py [hunk 3]
```diff
@@ -1,139 +1,118 @@
...
+          0.03382925, 0.0282597, 0.02383459, 0.01651026],
+         [0.0414435, 0.04603647, 0.05458152, 0.04969863, 0.03799496, 0.0325475,
+          0.02928206, 0.02685588, 0.02172885, 0.01722743],
+         [0.04107728, 0.04161135, 0.04798963, 0.05156023, 0.03952087, 0.02899214,
+          0.02589456, 0.02824444, 0.01815823, 0.01557945],
+         [0.03901731, 0.03302052, 0.03498893, 0.03929199, 0.03695735, 0.02943466,
+          0.01945525, 0.01869231, 0.01666284, 0.01240558],
+         [0.02664226, 0.02386511, 0.02206454, 0.02978561, 0.03265431, 0.0265507,
+          0.02214084, 0.01844815, 0.01542687, 0.01353475]],
+        dtype=np.float32
```
Comment (ttung) [inline starfish/test/full_pipelines/api/test_iss_api.py]: is it worth asserting on the dtype of `white_top_hat_filtered_image`?

File: starfish/test/full_pipelines/api/test_dartfish.py [hunk 4]
```diff
@@ -1,26 +1,24 @@
...
+dartfish = __import__('DARTFISH_Pipeline_-_Human_Occipital_Cortex_-_1_FOV')
+
 
 def test_dartfish_pipeline_cropped_data():
 
     # set random seed to errors provoked by optimization functions
     np.random.seed(777)
```
Comment (ttung) [inline starfish/test/full_pipelines/api/test_dartfish.py]: I believe you need to set the random seed before you run the pipeline.

Consider just moving the __import__ statement into this method.

File: starfish/test/full_pipelines/api/test_iss_api.py [hunk 5]
```diff
@@ -1,139 +1,118 @@
...
+
+os.environ["USE_TEST_DATA"] = "1"
+sys.path.append(os.path.join(ROOT_DIR, "notebooks/py/"))
+iss = __import__('ISS_Pipeline_-_Breast_-_1_FOV')
 
 
 def test_iss_pipeline_cropped_data():
 
     # set random seed to errors provoked by optimization functions
     np.random.seed(777)
```
Comment (ttung) [inline starfish/test/full_pipelines/api/test_iss_api.py]: ditto this random seeding.

File: starfish/test/full_pipelines/api/test_merfish.py [hunk 6]
```diff
@@ -1,25 +1,23 @@
...
-from starfish.types import Features, Indices
+os.environ["USE_TEST_DATA"] = "1"
+sys.path.append(os.path.join(ROOT_DIR, "notebooks/py/"))
+merfish = __import__('MERFISH_Pipeline_-_U2O2_Cell_Culture_-_1_FOV')
 
 
 def test_merfish_pipeline_cropped_data():
 
     # set random seed to errors provoked by optimization functions
     np.random.seed(777)
```
Comment (ttung) [inline starfish/test/full_pipelines/api/test_merfish.py]: ditto this random seed.

File: starfish/test/full_pipelines/api/test_merfish.py [hunk 7]
```diff
@@ -137,113 +130,92 @@ def test_merfish_pipeline_cropped_data():
...
-        max_area=np.inf,
-        norm_order=2,
-        crop_size=(0, 40, 40)
-    )
-    spot_intensities, prop_results = psd.run(low_passed)
+    spot_intensities = merfish.initial_spot_intensities
 
     # verify that the number of spots are correct
     spots_passing_filters = spot_intensities[Features.PASSES_THRESHOLDS].sum()
-    assert spots_passing_filters == 1125
```
Comment (ttung) [inline starfish/test/full_pipelines/api/test_merfish.py]: this is a significant change.  expected?  is this because of the magic scale factors thing?

File: notebooks/py/DARTFISH_Pipeline_-_Human_Occipital_Cortex_-_1_FOV.py [hunk 8]
```diff
@@ -271,17 +272,20 @@ def compute_magnitudes(stack, norm_order=2):
...
-
-plt.plot(real_barcode, 'ok')
-plt.stem(read_out_barcode)
-sns.despine(offset=2)
-plt.xticks(range(18))
-plt.title(gene)
-plt.xlabel('Index into R (0:5) and C(0:2)');
+# The test will error here on pixel_traces[ind,:] with an out of index error
+# because we are using the test data.
+if not test:
```
Comment (ambrosejcarr) [inline notebooks/py/DARTFISH_Pipeline_-_Human_Occipital_Cortex_-_1_FOV.py]: Is there an ind that we can pick that works for both data types? Can you try some smaller numbers? The selected Ind is not important.

File: notebooks/py/ISS_Pipeline_-_Breast_-_1_FOV.py [hunk 9]
```diff
@@ -220,12 +211,6 @@
 table = pd.Series(counts, index=genes).sort_values(ascending=False)
```
Comment (ambrosejcarr) [inline notebooks/py/ISS_Pipeline_-_Breast_-_1_FOV.py]: Add the table.head() back.

File: notebooks/py/MERFISH_Pipeline_-_U2O2_Cell_Culture_-_1_FOV.py [hunk 10]
```diff
@@ -114,7 +119,7 @@
 # EPY: START code
 scale_factors = {
     (t[Indices.ROUND], t[Indices.CH]): t['scale_factor']
-    for index, t in primary_image.tile_metadata.iterrows()
+    for t in experiment.extras['scale_factors']
```
Comment (ambrosejcarr) [inline notebooks/py/MERFISH_Pipeline_-_U2O2_Cell_Culture_-_1_FOV.py]: Nice.

File: starfish/test/full_pipelines/api/test_iss_api.py [hunk 11]
```diff
@@ -1,139 +1,123 @@
...
-    gsd = GaussianSpotDetector(
-        min_sigma=1,
-        max_sigma=10,
-        num_sigma=30,
-        threshold=0.01,
-        measurement_type='mean',
+    registered_image = iss.registered_image
+
+    # assert on a random part of the filtered image
+    expected_registered_values = np.array(
```
Comment (ambrosejcarr) [inline starfish/test/full_pipelines/api/test_iss_api.py]: Did the values here and for the other image assertion change?

File: starfish/test/full_pipelines/api/test_iss_api.py [hunk 12]
```diff
@@ -1,139 +1,123 @@
...
-
-    # segment
-    # TODO ambrosejcarr: do these need to be adjusted for the image size?
-    seg = Watershed(
-        dapi_threshold=.16,
-        input_threshold=.22,
-        min_distance=57
-    )
-    regions = seg.run(filtered_primary_image, filtered_nuclei)
+    genes, gene_counts = iss.genes, iss.counts
```
Comment (ambrosejcarr) [inline starfish/test/full_pipelines/api/test_iss_api.py]: I think the way these get calculated, the counts come out sorted. I'm wondering why these are not in sorted order.

File: starfish/test/full_pipelines/api/test_iss_api.py [hunk 13]
```diff
@@ -1,139 +1,123 @@
...
+         [0.04107728, 0.04161135, 0.04798963, 0.05156023, 0.03952087, 0.02899214,
+          0.02589456, 0.02824444, 0.01815823, 0.01557945],
+         [0.03901731, 0.03302052, 0.03498893, 0.03929199, 0.03695735, 0.02943466,
+          0.01945525, 0.01869231, 0.01666284, 0.01240558],
+         [0.02664226, 0.02386511, 0.02206454, 0.02978561, 0.03265431, 0.0265507,
+          0.02214084, 0.01844815, 0.01542687, 0.01353475]],
+        dtype=np.float32
     )
+
+    assert (white_top_hat_filtered_image.numpy_array.dtype == np.float32)
```
Comment (ttung) [inline starfish/test/full_pipelines/api/test_iss_api.py]: remove parens.

File: starfish/test/full_pipelines/api/test_dartfish.py [hunk 14]
```diff
@@ -56,13 +56,14 @@ def test_dartfish_pipeline_cropped_data():
         dtype=np.float32
     )
 
+    assert (primary_image.numpy_array.dtype == np.float32)
```
Comment (ttung) [inline starfish/test/full_pipelines/api/test_dartfish.py]: remove parens

File: starfish/test/full_pipelines/api/test_dartfish.py [hunk 15]
```diff
@@ -87,14 +88,14 @@ def test_dartfish_pipeline_cropped_data():
           0.01960784, 0.01960784, 0., 0., 0.]],
         dtype=np.float32,
     )
+    assert (normalized_image.numpy_array.dtype == np.float32)
```
Comment (ttung) [inline starfish/test/full_pipelines/api/test_dartfish.py]: remove parens
